crates/ide-assists/src/handlers/toggle_async_sugar.rs

```diff
@@ -0,0 +1,460 @@
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    famous_defs::FamousDefs,
+};
+use syntax::{
+    ast::{self, HasVisibility},
+    AstNode, NodeOrToken, SyntaxKind, SyntaxNode, SyntaxToken, TextRange,
+};
+
+use crate::{AssistContext, Assists};
+
+// Assist: toggle_async_sugar
+//
+// Rewrites asynchronous function into `impl Future` and back.
+// This action does not touch the function body and therefore `async { 0 }`
+// block does not transform to just `0`.
+//
+// ```
+// pub async f$0n foo() -> usize {
+//     0
+// }
+// ```
+// ->
+// ```
+// pub fn foo() -> impl Future<Output = usize> {
+//     0
+// }
+// ```
+pub(crate) fn toggle_async_sugar(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let function: ast::Fn = ctx.find_node_at_offset()?;
```

Comment:
Let's instead look for an `async` token or if that fails for a return type node, and then check if the parent of either is a function and continue as written. Reason being we don't want to show this assist whenever the cursor is in any part of the function as that clutters things.

Reply:
Changed, perhaps `unnecessary_async` assist should also be restricted